homeassistant/components/tasmota/device_trigger.py

```diff
@@ -0,0 +1,287 @@
+        vol.Required(CONF_PLATFORM): DEVICE,
+        vol.Required(CONF_DOMAIN): DOMAIN,
+        vol.Required(CONF_DEVICE_ID): str,
+        vol.Required(CONF_DISCOVERY_ID): str,
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_SUBTYPE): cv.string,
+    }
+)
+
+DEVICE_TRIGGERS = "tasmota_device_triggers"
+
+
+@attr.s(slots=True)
+class TriggerInstance:
+    """Attached trigger settings."""
+
+    action: AutomationActionType = attr.ib()
+    automation_info: dict = attr.ib()
+    trigger: "Trigger" = attr.ib()
+    remove: Optional[CALLBACK_TYPE] = attr.ib(default=None)
+
+    async def async_attach_trigger(self):
+        """Attach event trigger."""
+        event_config = {
+            event_trigger.CONF_PLATFORM: "event",
+            event_trigger.CONF_EVENT_TYPE: TASMOTA_EVENT,
+            event_trigger.CONF_EVENT_DATA: {
+                "mac": self.trigger.tasmota_trigger.cfg.mac,
+                "source": self.trigger.tasmota_trigger.cfg.source,
+                "event": self.trigger.tasmota_trigger.cfg.event,
+            },
+        }
+
+        event_config = event_trigger.TRIGGER_SCHEMA(event_config)
+        if self.remove:
+            self.remove()
+        self.remove = await event_trigger.async_attach_trigger(
+            self.trigger.hass,
+            event_config,
+            self.action,
+            self.automation_info,
+            platform_type="device",
+        )
+
+
+@attr.s(slots=True)
+class Trigger:
+    """Device trigger settings."""
+
+    device_id: str = attr.ib()
+    discovery_hash: dict = attr.ib()
+    hass: HomeAssistantType = attr.ib()
+    remove_update_signal: Callable[[], None] = attr.ib()
+    subtype: str = attr.ib()
+    tasmota_trigger: TasmotaTrigger = attr.ib()
+    type: str = attr.ib()
+    trigger_instances: List[TriggerInstance] = attr.ib(factory=list)
+
+    async def add_trigger(self, action, automation_info):
+        """Add Tasmota trigger."""
+        instance = TriggerInstance(action, automation_info, self)
+        self.trigger_instances.append(instance)
+
+        if self.tasmota_trigger is not None:
+            # If we know about the trigger, set it up
+            await instance.async_attach_trigger()
+
+        @callback
+        def async_remove() -> None:
+            """Remove trigger."""
+            if instance not in self.trigger_instances:
+                raise HomeAssistantError("Can't remove trigger twice")
+
+            if instance.remove:
+                instance.remove()
+            self.trigger_instances.remove(instance)
+
+        return async_remove
+
+    def detach_trigger(self):
+        """Remove Tasmota device trigger."""
+        # Mark trigger as unknown
+        self.tasmota_trigger = None
+
+        # Unsubscribe if this trigger is in use
+        for trig in self.trigger_instances:
+            if trig.remove:
+                trig.remove()
+                trig.remove = None
+
+    async def arm_tasmota_trigger(self):
```

Comment:
arm or activate ?

Reply:
I think it makes sense to "arm" a trigger :)

Reply:
We've been using "attach" in `helpers/trigger.py`

Reply:
Yes, but there's already `attach` going on. I have a hard time reading the code myself, there are the triggers defined by the Tasmota device, the triggers defined by the user in automations and the underlying event triggers.

Any suggestion for how to make it more readable? Maybe more comments?